AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/holder/ReceiverRegistrationHolder.java

```diff
@@ -58,17 +58,25 @@ private JFieldVar createIntentFilterField(IntentFilterData intentFilterData) {
 		JExpression newIntentFilterExpr = _new(classes().INTENT_FILTER);
 		JFieldVar intentFilterField = getGeneratedClass().field(PRIVATE | FINAL, classes().INTENT_FILTER, intentFilterName, newIntentFilterExpr);
 
-		JBlock initBody = holder.getInitBody();
+		JBlock intentFilterTarget = getIntentFilterTarget(intentFilterData);
 		for (String action : intentFilterData.getActionSet()) {
-			initBody.invoke(intentFilterField, "addAction").arg(action);
+			intentFilterTarget.invoke(intentFilterField, "addAction").arg(action);
 		}
 		for (String dataScheme : intentFilterData.getDataSchemeSet()) {
-			initBody.invoke(intentFilterField, "addDataScheme").arg(dataScheme);
+			intentFilterTarget.invoke(intentFilterField, "addDataScheme").arg(dataScheme);
 		}
 
 		return intentFilterField;
 	}
 
+	private JBlock getIntentFilterTarget(IntentFilterData intentFilterData) {
+		if (intentFilterData.registerAt.equals(RegisterAt.OnAttachOnDetach) && holder instanceof EFragmentHolder) {
```

Comment:
I think using polymorphism and method overrides is a better design. Maybe you could create this method in `EComponentHolder` which just returns init body, and override it in `EFragmentHolder` to return the needed block?